File: 2025/day1.py
import re

# with open("day1.txt", "r") as file:
with open("day1-sample.txt", "r") as file:
    p1_count = 0
    p2_count = 0
    position = 50
    for line in file:
        match = re.search(r"([LR])(\d+)", line)
        if match:
            direction = match.group(1)
            distance = int(match.group(2))

            orig_position = position

            if direction == "L":
                position -= distance
            elif direction == "R":
                position += distance

            passes = abs(position) // 100
            p2_count += passes 

            if position < 0 or position > 99:
                position = position % 100

            if position == 0:
                p1_count += 1

            print("move", line.strip())
            print("position", position)
            print() 

    print("part 1 count", p1_count)
    print("part 2 count", p1_count + p2_count)

File: 2025/day1.cs
var file = args switch
{
    [var path] => path,
    _ => Console.ReadLine() ?? throw new ArgumentException("No file path provided")
};

var lines = File.ReadAllLines(file);

Console.WriteLine($"Part 1: {Dial(Part1(lines)).Count(x => x == 0)}");
Console.WriteLine($"Part 2: {Dial(Part2(lines)).Count(x => x == 0)}");

static IEnumerable<int> Dial(IEnumerable<int> rotations)
{
    int position = 50;
    foreach (var rotation in rotations)
    {
        position = ((position + rotation) % 100 + 100) % 100;
        yield return position;
    }
}

static IEnumerable<int> Part1(IEnumerable<string> lines)
{
    return lines.Select(line =>
    {
        var direction = line[0];
        var distance = int.Parse(line[1..]);
        return (direction == 'R' ? 1 : -1) * distance;
    });
}

static IEnumerable<int> Part2(IEnumerable<string> lines)
{
    return lines.SelectMany(line =>
    {
        var direction = line[0];
        var distance = int.Parse(line[1..]);
        return Enumerable.Range(0, distance).Select(x => direction == 'R' ? 1 : -1);
    });
}
